### PR TITLE
Fix #4792 - autosized methods

### LINKED ISSUE #4792: Add Autosizing/Hardsizing Methods from OpenStudio-Standards
Enhancement Request
--------------
Implement autosize/hardsize-related methods from OpenStudio-Standards that have proven useful but which have not yet been implemented in C++.

## Detailed Description
The autosizing methods below fall into several categories. 
1. They are calculated using a combination of other autosizable fields, for example `Coil:Cooling:Water.autosizedDesignCoilLoad()`
2. They pull sizing values from an associated object, for example, some values of a `ThermalZone` actually come from the `Sizing:Zone` object associated with it.
3. They are utility methods that have proven useful and should be in C++.

The Ruby implementations of these methods is available in the openstudio-standards files linked below. Some of the fuel-related methods should probably be implemented on the ModelObject classes instead of on Model.

## Possible Implementation

[Siz.AirLoopHVAC.rb](https://github.com/NREL/openstudio-standards/blob/ba5168e63a87ac25d6092a61658f0741a0f8e2e8/lib/openstudio-standards/hvac_sizing/Siz.AirLoopHVAC.rb)
```
OpenStudio::Model::AirLoopHVAC

  autosizedSumMinimumHeatingAirFlowRates()

  autosizedSumAirTerminalMaxAirFlowRate()
```

[Siz.CoilCoolingWater.rb](https://github.com/NREL/openstudio-standards/blob/ba5168e63a87ac25d6092a61658f0741a0f8e2e8/lib/openstudio-standards/hvac_sizing/Siz.CoilCoolingWater.rb)
```
OpenStudio::Model::CoilCoolingWater
  
  autosizedDesignCoilLoad()
```

[Siz.ThermalZone.rb](https://github.com/NREL/openstudio-standards/blob/ba5168e63a87ac25d6092a61658f0741a0f8e2e8/lib/openstudio-standards/hvac_sizing/Siz.ThermalZone.rb)
```
OpenStudio::Model::ThermalZone

  autosizedMaximumOutdoorAirFlowRate()

  autosizedMinimumOutdoorAirFlowRate()

  autosizedCoolingDesignAirFlowRate()
  
  autosizedHeatingDesignAirFlowRate()
  
  coolingDesignLoad()
  
  designAirFlowRate()
  
  heatingDesignLoad()
  
  heating_fuels()
  
  cooling_fuels()
```

[Siz.HeatingCoolingFuels.rb](https://github.com/NREL/openstudio-standards/blob/ba5168e63a87ac25d6092a61658f0741a0f8e2e8/lib/openstudio-standards/hvac_sizing/Siz.HeatingCoolingFuels.rb) 
```
OpenStudio::Model::Model

  plant_loop_heating_fuels()
  
  plant_loop_cooling_fuels()
  
  coil_heating_fuels()
  
  coil_heat_type()
  
  airloop_primary_heat_type()
  
  zone_equipment_heat_type()
  
  coil_cooling_fuels()
  
  zone_equipment_heating_fuels()
  
  zone_equipment_cooling_fuels()
  
  zone_airloop_heating_fuels()
  
  zone_airloop_cooling_fuels()
```

Also add all the Ruby methods defined in these files that aren't already in OpenStudio C++, which will enable a lot of code and files to be removed from openstudio-results measure, as pointed out by @eringold.

https://github.com/NREL/openstudio-common-measures-gem/tree/develop/lib/measures/openstudio_results/resources

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes #4792
- Companion PR: https://github.com/NREL/OpenStudio-resources/pull/189
- At 00d74a51e6cd548d26fad3e4f81480448fe3a876, this is just refactoring to use ComponentSizes instead

### Pull Request Author

 - [x] Model API Changes / Additions
 - [x] Model API methods are tested: https://github.com/NREL/OpenStudio-resources/pull/189
 - [x] All new and existing tests passes

**Labels:**

 - [x] If change to an IDD file, add the label `IDDChange`
 - [x] If breaking existing API, add the label `APIChange`
 - [x] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: src/model/ModelObject_Impl.hpp [hunk 1]
```diff
@@ -221,11 +221,15 @@ namespace model {
...
-     *  schedule's ScheduleTypeLimits have not yet been set (in which case the ScheduleTypeRegistry
-     *  is used to retrieve or create an appropriate one). */
+       *  ScheduleType in the ScheduleTypeRegistry for (className,scheduleDisplayName), or if
+       *  schedule's ScheduleTypeLimits have not yet been set (in which case the ScheduleTypeRegistry
+       *  is used to retrieve or create an appropriate one). */
       bool setSchedule(unsigned index, const std::string& className, const std::string& scheduleDisplayName, Schedule& schedule);
 
+      /** For stuff that's plain missing from ComponentSizes table in E+, so getAutosizedValue can't work.
+        * This method is slower as it executes a lot more queries (an average of 10, versus 1 for getAutosizedValue) */
+      boost::optional<double> getAutosizedValueFromInitializationSummary(std::string valueName, std::string units) const;
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/ModelObject_Impl.hpp:231]: I actually need to for a couple of things

File: src/model/ThermalZone.hpp [hunk 2]
```diff
@@ -541,6 +541,17 @@ namespace model {
...
+    // SQL Queries
+    boost::optional<double> getAutosizedValueFromZoneSizes(const std::string& columnName, const std::string& loadType) const;
+
+    boost::optional<double> autosizedMaximumOutdoorAirFlowRate() const;
+    boost::optional<double> autosizedMinimumOutdoorAirFlowRate() const;
+    boost::optional<double> autosizedCoolingDesignAirFlowRate() const;
+    boost::optional<double> autosizedHeatingDesignAirFlowRate() const;
+    boost::optional<double> coolingDesignLoad() const;
+    boost::optional<double> designAirFlowRate() const;
+    boost::optional<double> heatingDesignLoad() const;
```
Comment (jmarrec) [inline src/model/ThermalZone.hpp:553]: I'm not sold on the naming. `autosizedCoolingDesignLoad` would be a bit weird too probably.
Has someone got an idea to clearly convey the idea that this is a grab from the sql?

@asparke2 @tijcolem @kbenne @joseph-robertson

File: src/model/ThermalZone.cpp [hunk 3]
```diff
@@ -2649,6 +2651,81 @@ namespace model {
...
+      boost::to_upper(sqlName);
+
+      const std::string directQuery = fmt::format(
+        R"sql(
+SELECT {} FROM ZoneSizes
+  WHERE ZoneName = ?
+    AND LoadType = ?;
+      )sql",
+        columnName);
+      boost::optional<double> val = model().sqlFile().get().execAndReturnFirstDouble(directQuery,
```
Comment (jmarrec) [inline src/model/ThermalZone.cpp:2679]: Query ZoneSizes

File: src/model/ThermalZone.cpp [hunk 4]
```diff
@@ -2649,6 +2651,81 @@ namespace model {
...
+      }
+
+      return val;
+    }
+
+    // SQL Queries
+    boost::optional<double> ThermalZone_Impl::autosizedMaximumOutdoorAirFlowRate() const {
+      //TODO: I don't think this is available anywhere actually
+      return boost::none;
+    }
```
Comment (jmarrec) [inline src/model/ThermalZone.cpp:2697]: * [ ] TODO: I can't find this one anywhere.
 
@asparke2 ideas please?

**Comment by tijcolem** [comment]
@jmarrec Ignore the window unit test failure for now.  I guess this somehow got introduced with the perf pull request a few days ago. I looked into it briefly but not sure what's going on.  Regardless, it's unrelated to this.

File: src/model/ThermalZone.cpp [hunk 5]
```diff
@@ -2631,6 +2633,81 @@ namespace model {
...
+
+    boost::optional<double> ThermalZone_Impl::heatingDesignLoad() const {
+      return getAutosizedValueFromZoneSizes("UserDesLoad", "Heating");
+    }
+
+    boost::optional<double> ThermalZone_Impl::designAirFlowRate() const {
+      auto coolingFlow_ = autosizedCoolingDesignAirFlowRate();
+      auto heatingFlow_ = autosizedHeatingDesignAirFlowRate();
+      if (coolingFlow_ && heatingFlow_) {
+        return coolingFlow_.get() + heatingFlow_.get();
```
Comment (mdahlhausen) [inline src/model/ThermalZone.cpp:2706]: I think this should be a maximum of the two, not a sum.  This is what we have in openstudio-standards:
https://github.com/NREL/openstudio-standards/blob/master/lib/openstudio-standards/hvac_sizing/Siz.ThermalZone.rb#L223-L224

```
# Use maximum of the two as actual design flow rate
result = OpenStudio::OptionalDouble.new([clg_des_air_flow_rate.to_f, htg_des_air_flow_rate.to_f].max)
```

File: src/model/ModelObject.cpp [hunk 6]
```diff
@@ -225,92 +226,28 @@ namespace model {
...
-      // Special logic to deal with EnergyPlus inconsistencies
-      if (sqlObjectType == "Coil:Heating:Gas") {
-        sqlObjectType = "Coil:Heating:Fuel";
-      }
-
-      if (sqlObjectType == "CoilPerformance:DX:Cooling") {
-        // Get the parent object
-        boost::optional<CoilCoolingDXTwoStageWithHumidityControlMode> parentCoil;
-        auto coilTwoSpdHumCtrls = this->model().getConcreteModelObjects<CoilCoolingDXTwoStageWithHumidityControlMode>();
-        for (const auto& coilInModel : coilTwoSpdHumCtrls) {
```
[the remark this answers is not in the document]
Comment (asparke2) [reply src/model/ModelObject.cpp:257]: perfect!

File: src/model/ThermalZone.cpp (hunk 4 above)
Comment (asparke2) [reply src/model/ThermalZone.cpp:2679]: @jmarrec I also don't see this used in openstudio-standards code or a value in the sql file I looked in, probably should remove the method.

File: src/model/ThermalZone.cpp (hunk 5 above)
Comment (asparke2) [reply src/model/ThermalZone.cpp:2706]: Agree, definitely not the sum.

File: src/model/ThermalZone.hpp (hunk 2 above)
Comment (asparke2) [reply src/model/ThermalZone.hpp:558]: I think that autosizedCoolingDesignLoad, etc. is good. Any thing that is "autosizedFoo" comes from the SQL file

**Review by asparke2 — CHANGES_REQUESTED** [review]
Change the names to autosizedDesignCoolingLoad, etc. and I think this looks good!

File: src/model/ThermalZone.cpp (hunk 4 above)
Comment (jmarrec) [reply src/model/ThermalZone.cpp:2679]: Removed in 981d0baa3062eaa1fcf10c700c6f04b81febd698

File: src/model/ThermalZone.hpp (hunk 2 above)
Comment (jmarrec) [reply src/model/ThermalZone.hpp]: Renamed in 4f7cd195ee54b56ab0c4070d670f0d307c46435d

File: src/model/ThermalZone.cpp (hunk 5 above)
Comment (jmarrec) [reply src/model/ThermalZone.cpp]: Done in 66dbd48c98bbc67897051301a7d85a988faa6484 , thanks for the catch!

**Pull request merged**